doc/run-cloud-build-locally.md

```diff
@@ -0,0 +1,197 @@
+to the FHIR server that you brought up using the
+[Synthea data uploader](https://github.com/google/fhir-data-pipes/blob/master/synthea-hiv/README.md#Uploader).
+
+The uploader requires that you install the `uploader` module requirements first.
+You can do this by running:
+
+```bash
+pip3 install -r ./synthea-hiv/uploader/requirements.txt
+```
+
+Please note, it is a good idea to first create a Python `virtualenv` before
+running the above command to avoid conflicts with other Python packages you may
+have installed globally. You can do this by running:
+
+```bash
+$ virtualenv -p python3.8 venv
+$ . ./venv/bin/activate
+```
+
+If you had this already set up, just activate your existing virtualenv.
+
+```bash
+$ workon venv
+```
+
+Then, you can run the uploader script to upload the synthetic data to the FHIR
+server.
+
+For example, to upload to the HAPI FHIR server brought up in the previous step,
+run:
+
+```bash
+python3 ./synthea-hiv/uploader/main.py HAPI http://localhost:9001/fhir \
+--input_dir ./synthea-hiv/sample_data --cores 8
+```
+
+Depending on your machine, using too many cores may slow down your machine or
+cause JDBC connection pool errors with the HAPI FHIR server. Reducing the number
+of cores using the `--cores` flag should help at the cost of increasing the time
+to upload the data.
+
+### 4. Build the e2e controller/spark image
+
+```bash
+docker build -t ${_REPOSITORY}/e2e-tests/controller-spark:${_TAG} -f e2e-tests/controller-spark/Dockerfile .
+```
+
+### 5. Build and run the batch pipeline
+
+Use the Maven-built image (or rebuild with `pipelines/batch/docker` if you
+change it):
+
+```bash
+cd pipelines/batch
+docker build -t ${_REPOSITORY}/batch-pipeline:${_TAG} .
+cd ../..
+```
+
+Run the FHIR-search job against HAPI:
+
+```bash
+docker run --rm \
+  -e FHIR_SERVER_URL=http://localhost:9001/fhir \
+  -e PARQUET_PATH=/workspace/e2e-tests/FHIR_SEARCH_HAPI \
+  -e OUTPUT_PARQUET_VIEW_PATH=/workspace/e2e-tests/FHIR_SEARCH_HAPI/VIEWS_TIMESTAMP_1 \
+  -e SINK_PATH=http://localhost:9001/fhir \
+  -e SINK_USERNAME=hapi -e SINK_PASSWORD=hapi \
+  -v $(pwd):/workspace \
+  --network host  \
+  ${_REPOSITORY}/batch-pipeline:${_TAG}
+```
+
+Repeat equivalent runs for JDBC mode, Bulk Export mode, and OpenMRS flows by
+applying the same env vars/cloud build steps from `cloudbuild.yaml`.
+
+### 6. Run e2e tests
+
+Execute the HAPI search test:
+
+```bash
+  docker run --rm -e PARQUET_SUBDIR=FHIR_SEARCH_HAPI   \
+    -e FHIR_JSON_SUBDIR=FHIR_SEARCH_HAPI_JSON  \
+    -e SINK_SERVER=http://localhost:9001 \
+    -e DOCKER_NETWORK=host   \
+    -v $(pwd):/workspace   \
+    --network host    \
+    ${_REPOSITORY}/e2e-tests:${_TAG}
+```
+
+or just the script directly:
+
+```bash
+./e2e-tests/pipeline_validation.sh e2e-tests/ FHIR_SEARCH_HAPI FHIR_SEARCH_HAPI_JSON http://localhost:9001/fhir
+```
+
+Repeat for each mode (JDBC, Bulk Export, controller/spark, OpenMRS) using the
```

Comment:
ditto (regarding "if required")
Also, please separate the `controller-spark` image from the rest and add a separate section for it. Its execution and debugging is different.